File: binding/src/libqkeysequence.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qkeysequence.html
pub const qkeysequence = struct {

    /// New constructs a new QKeySequence object.
    pub fn New() QtC.QKeySequence {
        return qtc.QKeySequence_new();
    }


    /// New2 constructs a new QKeySequence object.
    pub fn New2(key: []const u8) QtC.QKeySequence {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QKeySequence_new2(key_str);
    }


    /// New3 constructs a new QKeySequence object.
    pub fn New3(k1: i32) QtC.QKeySequence {
        return qtc.QKeySequence_new3(k1);
    }


    /// New4 constructs a new QKeySequence object.
    pub fn New4(k1: QtC.QKeyCombination) QtC.QKeySequence {
        return qtc.QKeySequence_new4(@ptrCast(k1));
    }


    /// New5 constructs a new QKeySequence object.
    pub fn New5(ks: ?*anyopaque) QtC.QKeySequence {
        return qtc.QKeySequence_new5(@ptrCast(ks));
    }


    /// New6 constructs a new QKeySequence object.
    pub fn New6(key: i32) QtC.QKeySequence {
        return qtc.QKeySequence_new6(@intCast(key));
    }


    /// New7 constructs a new QKeySequence object.
    pub fn New7(key: []const u8, format: i32) QtC.QKeySequence {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QKeySequence_new7(key_str, @intCast(format));
    }


    /// New8 constructs a new QKeySequence object.
    pub fn New8(k1: i32, k2: i32) QtC.QKeySequence {
        return qtc.QKeySequence_new8(k1, k2);
    }


    /// New9 constructs a new QKeySequence object.
    pub fn New9(k1: i32, k2: i32, k3: i32) QtC.QKeySequence {
        return qtc.QKeySequence_new9(k1, k2, k3);
    }


    /// New10 constructs a new QKeySequence object.
    pub fn New10(k1: i32, k2: i32, k3: i32, k4: i32) QtC.QKeySequence {
        return qtc.QKeySequence_new10(k1, k2, k3, k4);
    }


    /// New11 constructs a new QKeySequence object.
    pub fn New11(k1: QtC.QKeyCombination, k2: QtC.QKeyCombination) QtC.QKeySequence {
        return qtc.QKeySequence_new11(@ptrCast(k1), @ptrCast(k2));
    }


    /// New12 constructs a new QKeySequence object.
    pub fn New12(k1: QtC.QKeyCombination, k2: QtC.QKeyCombination, k3: QtC.QKeyCombination) QtC.QKeySequence {
        return qtc.QKeySequence_new12(@ptrCast(k1), @ptrCast(k2), @ptrCast(k3));
    }


    /// New13 constructs a new QKeySequence object.
    pub fn New13(k1: QtC.QKeyCombination, k2: QtC.QKeyCombination, k3: QtC.QKeyCombination, k4: QtC.QKeyCombination) QtC.QKeySequence {
        return qtc.QKeySequence_new13(@ptrCast(k1), @ptrCast(k2), @ptrCast(k3), @ptrCast(k4));
    }


    pub fn Count(self: ?*anyopaque, ) i32 {
        return qtc.QKeySequence_Count(@ptrCast(self));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QKeySequence_IsEmpty(@ptrCast(self));
    }

    pub fn ToString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QKeySequence_ToString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qkeysequence.ToString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FromString(str: []const u8) QtC.QKeySequence {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QKeySequence_FromString(str_str);
    }

    pub fn ListFromString(str: []const u8) []const u8 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QKeySequence_ListFromString(str_str);
    }

    pub fn ListToString(list: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QKeySequence_ListToString(@ptrCast(list));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qkeysequence.ListToString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Matches(self: ?*anyopaque, seq: ?*anyopaque) i32 {
        return qtc.QKeySequence_Matches(@ptrCast(self), @ptrCast(seq));
    }

    pub fn Mnemonic(text: []const u8) QtC.QKeySequence {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QKeySequence_Mnemonic(text_str);
    }

    pub fn KeyBindings(key: i32) []const u8 {
        return qtc.QKeySequence_KeyBindings(@intCast(key));
    }

    pub fn OperatorQVariant(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QKeySequence_OperatorQVariant(@ptrCast(self));
    }

    pub fn OperatorSubscript(self: ?*anyopaque, i: u32) QtC.QKeyCombination {
        return qtc.QKeySequence_OperatorSubscript(@ptrCast(self), i);
    }

    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QKeySequence_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QKeySequence_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorLesser(self: ?*anyopaque, ks: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorLesser(@ptrCast(self), @ptrCast(ks));
    }

    pub fn OperatorGreater(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorGreater(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorLesserOrEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorLesserOrEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorGreaterOrEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QKeySequence_OperatorGreaterOrEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsDetached(self: ?*anyopaque, ) bool {
        return qtc.QKeySequence_IsDetached(@ptrCast(self));
    }

    pub fn ToString1(self: ?*anyopaque, format: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QKeySequence_ToString1(@ptrCast(self), @intCast(format));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qkeysequence.ToString1: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FromString2(str: []const u8, format: i32) QtC.QKeySequence {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QKeySequence_FromString2(str_str, @intCast(format));
    }

    pub fn ListFromString2(str: []const u8, format: i32) []const u8 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QKeySequence_ListFromString2(str_str, @intCast(format));
    }

    pub fn ListToString2(list: []const u8, format: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QKeySequence_ListToString2(@ptrCast(list), @intCast(format));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qkeysequence.ListToString2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QKeySequence_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/keysequence.html#types
pub const enums = struct {
    pub const StandardKey = enum {
        pub const UnknownKey: i32 = 0;
        pub const HelpContents: i32 = 1;
        pub const WhatsThis: i32 = 2;
        pub const Open: i32 = 3;
        pub const Close: i32 = 4;
        pub const Save: i32 = 5;
        pub const New: i32 = 6;
        pub const Delete: i32 = 7;
        pub const Cut: i32 = 8;
        pub const Copy: i32 = 9;
        pub const Paste: i32 = 10;
        pub const Undo: i32 = 11;
        pub const Redo: i32 = 12;
        pub const Back: i32 = 13;
        pub const Forward: i32 = 14;
        pub const Refresh: i32 = 15;
        pub const ZoomIn: i32 = 16;
        pub const ZoomOut: i32 = 17;
        pub const Print: i32 = 18;
        pub const AddTab: i32 = 19;
        pub const NextChild: i32 = 20;
        pub const PreviousChild: i32 = 21;
        pub const Find: i32 = 22;
        pub const FindNext: i32 = 23;
        pub const FindPrevious: i32 = 24;
        pub const Replace: i32 = 25;
        pub const SelectAll: i32 = 26;
        pub const Bold: i32 = 27;
        pub const Italic: i32 = 28;
        pub const Underline: i32 = 29;
        pub const MoveToNextChar: i32 = 30;
        pub const MoveToPreviousChar: i32 = 31;
        pub const MoveToNextWord: i32 = 32;
        pub const MoveToPreviousWord: i32 = 33;
        pub const MoveToNextLine: i32 = 34;
        pub const MoveToPreviousLine: i32 = 35;
        pub const MoveToNextPage: i32 = 36;
        pub const MoveToPreviousPage: i32 = 37;
        pub const MoveToStartOfLine: i32 = 38;
        pub const MoveToEndOfLine: i32 = 39;
        pub const MoveToStartOfBlock: i32 = 40;
        pub const MoveToEndOfBlock: i32 = 41;
        pub const MoveToStartOfDocument: i32 = 42;
        pub const MoveToEndOfDocument: i32 = 43;
        pub const SelectNextChar: i32 = 44;
        pub const SelectPreviousChar: i32 = 45;
        pub const SelectNextWord: i32 = 46;
        pub const SelectPreviousWord: i32 = 47;
        pub const SelectNextLine: i32 = 48;
        pub const SelectPreviousLine: i32 = 49;
        pub const SelectNextPage: i32 = 50;
        pub const SelectPreviousPage: i32 = 51;
        pub const SelectStartOfLine: i32 = 52;
        pub const SelectEndOfLine: i32 = 53;
        pub const SelectStartOfBlock: i32 = 54;
        pub const SelectEndOfBlock: i32 = 55;
        pub const SelectStartOfDocument: i32 = 56;
        pub const SelectEndOfDocument: i32 = 57;
        pub const DeleteStartOfWord: i32 = 58;
        pub const DeleteEndOfWord: i32 = 59;
        pub const DeleteEndOfLine: i32 = 60;
        pub const InsertParagraphSeparator: i32 = 61;
        pub const InsertLineSeparator: i32 = 62;
        pub const SaveAs: i32 = 63;
        pub const Preferences: i32 = 64;
        pub const Quit: i32 = 65;
        pub const FullScreen: i32 = 66;
        pub const Deselect: i32 = 67;
        pub const DeleteCompleteLine: i32 = 68;
        pub const Backspace: i32 = 69;
        pub const Cancel: i32 = 70;
    };

    pub const SequenceFormat = enum {
        pub const NativeText: i32 = 0;
        pub const PortableText: i32 = 1;
    };

    pub const SequenceMatch = enum {
        pub const NoMatch: i32 = 0;
        pub const PartialMatch: i32 = 1;
        pub const ExactMatch: i32 = 2;
    };

};

File: binding/src/libqkeysequence.cpp
#include <QKeyCombination>
#include <QKeySequence>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QVariant>
#include <qkeysequence.h>
#include "libqkeysequence.h"
#include "libqkeysequence.hxx"

QKeySequence* QKeySequence_new() {
	 return new QKeySequence();
}

QKeySequence* QKeySequence_new2(const libqt_string key) {
	 return new QKeySequence(QString::fromUtf8(key.data, key.len));
}

QKeySequence* QKeySequence_new3(int k1) {
	 return new QKeySequence(k1);
}

QKeySequence* QKeySequence_new4(QKeyCombination* k1) {
	 return new QKeySequence(*k1);
}

QKeySequence* QKeySequence_new5(const QKeySequence* ks) {
	 return new QKeySequence(*ks);
}

QKeySequence* QKeySequence_new6(int key) {
	 return new QKeySequence(static_cast<QKeySequence::StandardKey>(key));
}

QKeySequence* QKeySequence_new7(const libqt_string key, int format) {
	 return new QKeySequence(QString::fromUtf8(key.data, key.len), static_cast<QKeySequence::SequenceFormat>(format));
}

QKeySequence* QKeySequence_new8(int k1, int k2) {
	 return new QKeySequence(k1, k2);
}

QKeySequence* QKeySequence_new9(int k1, int k2, int k3) {
	 return new QKeySequence(k1, k2, k3);
}

QKeySequence* QKeySequence_new10(int k1, int k2, int k3, int k4) {
	 return new QKeySequence(k1, k2, k3, k4);
}

QKeySequence* QKeySequence_new11(QKeyCombination* k1, QKeyCombination* k2) {
	 return new QKeySequence(*k1, *k2);
}

QKeySequence* QKeySequence_new12(QKeyCombination* k1, QKeyCombination* k2, QKeyCombination* k3) {
	 return new QKeySequence(*k1, *k2, *k3);
}

QKeySequence* QKeySequence_new13(QKeyCombination* k1, QKeyCombination* k2, QKeyCombination* k3, QKeyCombination* k4) {
	 return new QKeySequence(*k1, *k2, *k3, *k4);
}

int QKeySequence_Count(const QKeySequence* self) {
	return self->count();
}

bool QKeySequence_IsEmpty(const QKeySequence* self) {
	return self->isEmpty();
}

libqt_string QKeySequence_ToString(const QKeySequence* self) {
	QString _ret = self->toString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QKeySequence* QKeySequence_FromString(const libqt_string str) {
	return new QKeySequence(QKeySequence::fromString(QString::fromUtf8(str.data, str.len)));
}

libqt_list QKeySequence_ListFromString(const libqt_string str) {
	auto _ret = QKeySequence::listFromString(QString::fromUtf8(str.data, str.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_string QKeySequence_ListToString(const libqt_list list) {
	QString _ret = QKeySequence::listToString(QList<QKeySequence>());
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QKeySequence_Matches(const QKeySequence* self, const QKeySequence* seq) {
	return self->matches(*seq);
}

QKeySequence* QKeySequence_Mnemonic(const libqt_string text) {
	return new QKeySequence(QKeySequence::mnemonic(QString::fromUtf8(text.data, text.len)));
}

libqt_list QKeySequence_KeyBindings(int key) {
	auto _ret = QKeySequence::keyBindings(static_cast<QKeySequence::StandardKey>(key));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

QVariant* QKeySequence_OperatorQVariant(const QKeySequence* self) {
	return new QVariant(self->operator QVariant());
}

QKeyCombination* QKeySequence_OperatorSubscript(const QKeySequence* self, unsigned int i) {
	return new QKeyCombination(self->operator[](i));
}

void QKeySequence_OperatorAssign(QKeySequence* self, const QKeySequence* other) {
	self->operator=(*other);
}

void QKeySequence_Swap(QKeySequence* self, QKeySequence* other) {
	self->swap(*other);
}

bool QKeySequence_OperatorEqual(const QKeySequence* self, const QKeySequence* other) {
	return self->operator==(*other);
}

bool QKeySequence_OperatorNotEqual(const QKeySequence* self, const QKeySequence* other) {
	return self->operator!=(*other);
}

bool QKeySequence_OperatorLesser(const QKeySequence* self, const QKeySequence* ks) {
	return self->operator<(*ks);
}

bool QKeySequence_OperatorGreater(const QKeySequence* self, const QKeySequence* other) {
	return self->operator>(*other);
}

bool QKeySequence_OperatorLesserOrEqual(const QKeySequence* self, const QKeySequence* other) {
	return self->operator<=(*other);
}

bool QKeySequence_OperatorGreaterOrEqual(const QKeySequence* self, const QKeySequence* other) {
	return self->operator>=(*other);
}

bool QKeySequence_IsDetached(const QKeySequence* self) {
	return self->isDetached();
}

libqt_string QKeySequence_ToString1(const QKeySequence* self, int format) {
	QString _ret = self->toString(static_cast<QKeySequence::SequenceFormat>(format));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QKeySequence* QKeySequence_FromString2(const libqt_string str, int format) {
	return new QKeySequence(QKeySequence::fromString(QString::fromUtf8(str.data, str.len), static_cast<QKeySequence::SequenceFormat>(format)));
}

libqt_list QKeySequence_ListFromString2(const libqt_string str, int format) {
	auto _ret = QKeySequence::listFromString(QString::fromUtf8(str.data, str.len), static_cast<QKeySequence::SequenceFormat>(format));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_string QKeySequence_ListToString2(const libqt_list list, int format) {
	QString _ret = QKeySequence::listToString(QList<QKeySequence>(), static_cast<QKeySequence::SequenceFormat>(format));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QKeySequence_Delete(QKeySequence* self) {
    delete self;
}

